File: ThreadingSimulator/Models/ProgramModel.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Collections.Specialized;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;
using ThreadingSimulator.Other;

namespace ThreadingSimulator.Models
{
    [XmlInclude(typeof(ProcessBlocksModel))]
    [XmlInclude(typeof(BlockModel))]
    [XmlInclude(typeof(LoopBlockModel))]
    [XmlInclude(typeof(SemaphorBlockModel))]
    [XmlInclude(typeof(VariableBlockModel))]
    public class ProgramModel : PropertyChangedImpl
    {
        [XmlIgnore]
        [JsonIgnore]
        private bool subscribed = false;

        public string Name { get; set; }
        public ObservableCollection<ProcessBlocksModel> Processes { get; set; }

        [XmlIgnore]
        [JsonIgnore]
        public int ProcessCount
        {
            get
            {
                return Processes != null ? Processes.Count : 0;
            }
        }

        [XmlIgnore]
        [JsonIgnore]
        public bool IsValid
        {
            get
            {
                if (!String.IsNullOrEmpty(Name) && ProcessCount > 0 
                    && ProcessCount == Processes.Select(x=>x.Name).Distinct().Count())
                {
                    foreach(ProcessBlocksModel process in Processes)
                    {
                        if(!process.IsValid)
                        {
                            return false;
                        }
                    }

                    return true;
                }

                return false;
            }
        }
        
        public void Add(ProcessBlocksModel commands)
        {
            if(Processes==null)
            {
                Processes = new ObservableCollection<ProcessBlocksModel>();

                SubscribeToChanges();
                OnPropertyChanged(nameof(Processes));
                OnPropertyChanged(nameof(ProcessCount));
            }

            Processes.Add(commands);
        }

        private void Processes_CollectionChanged(object sender, NotifyCollectionChangedEventArgs e)
        {
            OnPropertyChanged(nameof(ProcessCount));
        }

        public void SubscribeToChanges()
        {
            if(Processes !=null && !subscribed)
            {
                Processes.CollectionChanged += Processes_CollectionChanged;
                subscribed = true;
            }
        }

        public ProgramModel Clone()
        {
            XmlSerializer xmlSerializer = new XmlSerializer(GetType());

            using (MemoryStream ms = new MemoryStream())
            {
                xmlSerializer.Serialize(ms, this);

                ms.Position = 0;

                ProgramModel returnModel = (ProgramModel)xmlSerializer.Deserialize(ms);
                returnModel.SubscribeToChanges();

                return returnModel;
            }
        }
    }
}

File: GUI/src/app/Models/ProgramModel.ts
import { ProcessBlocksModel } from "./ProcessBlocksModel";
import { isNullOrUndefined } from "util";

export class ProgramModel
{
    public Name: string;
    public Processes: ProcessBlocksModel[];

    public ProcessCount(): number
    {
        return this.Processes != null ? this.Processes.length : 0;
    }

    public IsValid(): boolean
    {
        if (!isNullOrUndefined(this.Name) && this.ProcessCount() > 0 
                && this.ProcessCount() == this.Distinct(this.Select(this.Processes, x=>x.Name)).length)
        {
            for(let i=0;i<this.ProcessCount();i++)
            {
                if(!this.Processes[i].IsValid())
                {
                    return false;
                }
            }

            return true;
        }

        return false;
    }
    
    public Add(commands: ProcessBlocksModel)
    {
        if(this.Processes==null)
        {
            this.Processes =  [];
        }

        this.Processes.push(commands);
    }

    public Clone(): ProgramModel
    {
        return ProgramModel.Deserialize(this);
    }

    public static Deserialize(obj: any): ProgramModel
    {
        let trueObj: ProgramModel = new ProgramModel();

        trueObj.Name = obj.Name;

        if(obj.Processes != null)
        {
            trueObj.Processes = [];
    
            obj.Processes.forEach(tmp => {
                trueObj.Processes.push(ProcessBlocksModel.Deserialize(tmp));
            });       
        }

        return trueObj;
    }

    public Select(arr: any[], func): any[]
    {
        let retArr = [];

        for(let obj in arr)
        {
            retArr.push(func(obj));
        }

        return retArr;
    }

    public Distinct(arr: any[]): any[]
    {
        let retArr = [];

        for(let obj in arr)
        {
            if(retArr.indexOf(obj) == -1)
            {
                retArr.push(obj);
            }
        }

        return retArr;
    }
}